sql/src/main/java/io/crate/expression/reference/sys/shard/ShardSegments.java

```diff
@@ -47,18 +49,24 @@ public ShardSegments(IndicesService indicesService) {
 
     @Override
     public Iterator<ShardSegment> iterator() {
-        List<ShardSegment> result = new ArrayList<>();
-        for (IndexService indexService : indicesService) {
-            for (IndexShard indexShard : indexService) {
-                ShardId shardId = indexShard.shardId();
-                ShardRouting shardRouting = indexShard.routingEntry();
-                if (shardId != null && !shardRouting.unassigned()) {
-                    for (Segment segment : indexShard.segments(false)) {
-                        result.add(new ShardSegment(shardId.id(), shardId.getIndexName(), segment, shardRouting.primary()));
-                    }
-                }
-            }
+        return stream(indicesService)
+            .flatMap(Streams::stream)
+            .filter(x -> !x.routingEntry().unassigned())
+            .flatMap(this::buildShardSegment)
+            .iterator();
+    }
+
+    private Stream<ShardSegment> buildShardSegment(IndexShard indexShard) {
+        try {
+            List<Segment> segments = indexShard.segments(false);
```

Comment:
This fixes the issues on the build pipeline, but I am not really happy with this solution. I am still surprised that the issues only occurs on jdk12.